src/libUtils/DnsUtils.cpp

```diff
@@ -0,0 +1,329 @@
+                               << url << " Err: " << errcode
+                               << ", Msg: " << gai_strerror(errcode)
+                               << ", retry: " << retry);
+      continue;
+    }
+    break;  // If we are here, means success
+  }
+
+  if (isShuttingDown) return false;
+
+  if (errcode != 0) {
+    LOG_GENERAL(WARNING, "Unable to connect " << url << ", Err: " << errcode);
+    return false;
+  }
+
+  res = result;
+
+  auto &ipStrList = dnsCacheListDataMap[listType].ipList;
+  ipStrList.clear();
+
+  while (res) {
+    switch (res->ai_family) {
+      case AF_INET:
+        ptr = &((struct sockaddr_in *)res->ai_addr)->sin_addr;
+        break;
+      case AF_INET6:
+        ptr = &((struct sockaddr_in6 *)res->ai_addr)->sin6_addr;
+        break;
+    }
+
+    if (ptr) {
+      inet_ntop(res->ai_family, ptr, addrstr, 100);
+      LOG_GENERAL(DEBUG, "Address: " << addrstr);
+      ipStrList.emplace_back(addrstr);
+    }
+    res = res->ai_next;
+    ptr = nullptr;
+  }
+
+  freeaddrinfo(result);
+
+  return true;
+}
+
+// Pub key are stored in the TXT record
+bool QueryPubKeyFromUrl(bytes &output, const string &pubKeyUrl) {
+  LOG_MARKER();
+
+  unsigned char query_buffer[256];
+  int resLen = 0;
+
+  for (unsigned int retry = 0; retry < QUERY_DNS_MAX_TRIES; ++retry) {
+    if (isShuttingDown) return false;
+
+    resLen = res_query(pubKeyUrl.c_str(), C_IN, ns_t_txt, query_buffer,
+                       sizeof(query_buffer));
+    if (resLen < 0) {
+      LOG_GENERAL(WARNING, "Failed to query pub key from "
+                               << pubKeyUrl << " Retry: " << retry);
+      continue;
+    }
+    break;  // If we are here, means success
+  }
+
+  if (isShuttingDown) return false;
+
+  if (resLen < 0) {
+    LOG_GENERAL(WARNING, "Unable to obtain pubkey from " << pubKeyUrl
+                                                         << " Res: " << resLen);
+    return false;
+  }
+
+  ns_msg nsMsg;
+  ns_initparse(query_buffer, resLen, &nsMsg);
+
+  if (ns_msg_count(nsMsg, ns_s_an) <= 0) {
+    LOG_GENERAL(WARNING, "No data found from pub key from " << pubKeyUrl);
+    return false;
+  }
+
+  ns_rr rr;
+  ns_parserr(&nsMsg, ns_s_an, 0, &rr);
+  auto rrData = ns_rr_rdata(rr) + 1;
+
+  string p(rrData, rrData + rr.rdlength - 1);
+
+  if (p.compare("0") == 0 || p.empty()) {
+    LOG_GENERAL(WARNING, "Pubkey from dns is 0 or empty: " << pubKeyUrl);
+    return false;
+  }
+
+  if (!DataConversion::HexStrToUint8Vec(p, output)) {
+    LOG_GENERAL(WARNING, "Invalid data obtained from pub key " << pubKeyUrl);
+    return false;
+  }
+  return true;
+}
+
+void QueryDnsList(DnsListType listType) {
+  LOG_MARKER();
+  if (isShuttingDown) return;
+
+  auto url = dnsAddresses[listType];
+  if (url.empty()) {
+    LOG_GENERAL(INFO, "DNS address is empty for type " << (int)listType);
+    return;
+  }
+
+  auto &dataListCache = dnsCacheListDataMap[listType];
+
+  if (!dataListCache.dataAccessMutex.try_lock()) {
```

Comment:
Just curious, are there situations where multiple threads may actually be doing query? Is it if two DS blocks are received (or mined) at somehow the same time?

In any case, good to have this in place.

Reply:
This mutex is trying to protect when querying is in progress.

As the query for `getaddrinfo` is a blocking call, sometimes it might take from 2ms to 15 seconds (So far when I observed querying fails).

In the chance that it blocks [indefinitely](https://man7.org/linux/man-pages/man3/getaddrinfo_a.3.html#:~:text=then%20the%20call-,blocks%20indefinitely,-(until%20one%0A%20%20%20%20%20%20%20%20%20%20of), at least this try_lock() will return false for those who are attempting to access the cache, so that we can continue to progress.